File: integrations/pqc-production/microbench.mjs
import assert from 'node:assert/strict';
import {createRequire} from 'node:module';
import {join,resolve} from 'node:path';
import {pathToFileURL} from 'node:url';
const require=createRequire(join(resolve(process.argv[2]),'package.json'));
const load=name=>import(pathToFileURL(require.resolve(name)));
const sdk=await load('@vellacognitive/vella-sdk');
const pq=await load('@vellacognitive/vella-sdk/pqc/index.js');
const keys=pq.generateHybridKeys(pq.SUITE),gov=sdk.createGovernor(undefined,{proofProfile:pq.HYBRID_PROFILE});
const proof=gov.govern({intent:'EXECUTE_CHANGE',evidenceMask:1,proof:{signingKey:{sign:r=>pq.signProofV3(r,keys.privateKeys,pq.SUITE)}}}).proofBundle;
const record=pq.verifyProofV3(proof,keys.publicKeys,pq.SUITE).authenticated;
function measure(fn,n=500){const samples=[];for(let i=0;i<n+25;i++){const start=performance.now();fn();if(i>=25)samples.push(performance.now()-start);}samples.sort((a,b)=>a-b);return {samples:n,p50:samples[Math.floor(.5*n)],p95:samples[Math.floor(.95*n)],p99:samples[Math.floor(.99*n)],max:samples.at(-1),mean:samples.reduce((a,b)=>a+b,0)/n};}
console.log(JSON.stringify({kind:'vella-node-v3-microbench',node:process.version,openssl:process.versions.openssl,units:'ms',keyBehavior:'Every sign imports both private keys; every verify imports both public keys. No cache.',evaluate:measure(()=>gov.govern({intent:'EXECUTE_CHANGE',evidenceMask:1})),sign:measure(()=>pq.signProofV3(record,keys.privateKeys,pq.SUITE)),verify:measure(()=>assert.equal(pq.verifyProofV3(proof,keys.publicKeys,pq.SUITE).ok,true)),proofBytes:Buffer.byteLength(JSON.stringify(proof)),scope:'Installed Node provider only; no MCP or storage in these timings'},null,2));

File: integrations/pqc-production/microbench.py
import json
import platform
import statistics
import time

import cryptography
from cryptography.hazmat.backends.openssl.backend import backend
from vella import create_governor
from vella.pqc import (
 SUITE,
 HybridProfile,
 generate_hybrid_keys,
 sign_proof_v3,
 verify_proof_v3,
)

pair=generate_hybrid_keys(SUITE); gov=create_governor(proof_profile=HybridProfile())
result=gov.govern(intent='EXECUTE_CHANGE',evidence_mask=1,proof_signing_key=pair['privateKeys'])
bundle=result['proof_bundle']; record=verify_proof_v3(bundle,pair['publicKeys'],SUITE)['authenticated']
def measure(fn,n=500):
 samples=[]
 for i in range(n+25):
  start=time.perf_counter_ns(); fn(); elapsed=(time.perf_counter_ns()-start)/1e6
  if i>=25:samples.append(elapsed)
 samples.sort()
 return {'samples':n,'p50':samples[int(.5*n)],'p95':samples[int(.95*n)],'p99':samples[int(.99*n)],'max':max(samples),'mean':statistics.mean(samples)}
def verify():
 assert verify_proof_v3(bundle,pair['publicKeys'],SUITE)['ok']
print(json.dumps({'kind':'vella-python-v3-microbench','python':platform.python_version(),'cryptography':cryptography.__version__,'openssl':backend.openssl_version_text(),'units':'ms','keyBehavior':'Every sign imports both private keys; every verify imports both public keys. No cache.','evaluate':measure(lambda:gov.govern(intent='EXECUTE_CHANGE',evidence_mask=1)),'sign':measure(lambda:sign_proof_v3(record,pair['privateKeys'],SUITE)),'verify':measure(verify),'proofBytes':len(json.dumps(bundle,separators=(',',':')).encode()),'scope':'Installed Python provider only; no MCP or storage in these timings'},indent=2))
